### PR TITLE
FastSmrLoader can set serializer

### PR DESCRIPTION
**Pull request by rogermichoud**
It was using the default serializer. Now it can take a serializer as a parameter.

### CODE REVIEW

File: test/src/test/java/org/corfudb/recovery/FastSmrMapsLoaderTest.java [hunk 1]
```diff
@@ -1173,7 +1175,32 @@ public void doNotReconstructEmptyCheckpoints() throws Exception {
         assertThat(streamTails.get(stream1)).isEqualTo(tail1);
         assertThat(streamTails.get(stream2)).isEqualTo(tail2);
         assertThat(streamTails.get(emptyStream)).isNull();
+    }
+
+    @Test
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/recovery/FastSmrMapsLoaderTest.java]: comments.

File: test/src/test/java/org/corfudb/recovery/FastSmrMapsLoaderTest.java [hunk 2]
```diff
@@ -1173,7 +1175,32 @@ public void doNotReconstructEmptyCheckpoints() throws Exception {
...
+
+        CorfuRuntime rt2 = new CorfuRuntime(getDefaultConfigurationString())
+                .connect();
+
+        FastSmrMapsLoader fsmr = new FastSmrMapsLoader(rt2);
+        fsmr.setSerializer(customSerializer);
+        fsmr.loadMaps();
 
+        test.get("a");
+        assertThat(test.get("a")).isEqualTo("b");
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/recovery/FastSmrMapsLoaderTest.java]: no negative test ?

File: test/src/test/java/org/corfudb/recovery/FastSmrMapsLoaderTest.java [hunk 3]
```diff
@@ -1173,7 +1175,32 @@ public void doNotReconstructEmptyCheckpoints() throws Exception {
         assertThat(streamTails.get(stream1)).isEqualTo(tail1);
         assertThat(streamTails.get(stream2)).isEqualTo(tail2);
         assertThat(streamTails.get(emptyStream)).isNull();
+    }
+
+    @Test
+    public void canUseCustomSerializer() throws Exception {
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/recovery/FastSmrMapsLoaderTest.java]: This test is really superfluous, using a custom serializer with objects view is already tested.
This test only tests that you can invoke a setter on the FastSmrMapsLoader object.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **8b4d132**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #887 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/s1O9h8EcfkKD80qLsJmS3M0YdAK6n6Gt)

**Comment by rogermichoud** [comment]
I changed the approach, setting a serializer for all the map was not the correct approach anyway (works only if all the maps have the same serializer).
Now when we create a new map, we take the entry serializer type to create the correct serializer.

Also added warning if we attempt to open an already existing map with a different serializer.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **e42a237**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #887 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/UN0sqPuK8d0JojkwZdk4BU9uL1qVQVhr)

**Comment by rogermichoud** [comment]
@medhavidhawan 
I updated the PR and added some comments for tests.

@Maithem 
replace the superfluous test with something that makes more sense. Also change the approach, you might want to take a look

File: test/src/test/java/org/corfudb/runtime/object/CorfuSMRObjectProxyTest.java [hunk 4]
```diff
@@ -260,4 +261,40 @@ public void canUseCustomSerializer() throws Exception {
...
     }
 
+    /**
+     * Once a SMR map is created with a serializer, we should
+     * not be able to change the serializer by re-opening the map with
+     * a different serializer.
+     *
+     * @throws Exception
+     */
+    @Test
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/runtime/object/CorfuSMRObjectProxyTest.java]: What about the case where the serializer is set from the persisted data and not the open argument?

**Pull request merged**